### ISSUE TITLE
use correct log level to make log better

### ISSUE THREAD

**Issue by Wing924** [netdata/netdata#5480]
##### Feature idea summary

Currently, we have too many error logs because we log `Something that might disable a part of netdata` as error. For example:
```
2019-02-23 14:56:42: python.d ERROR: web_log: nginx_log: /var/log/nginx/access.log not readable or not exist
```
If servers don't have nginx, it's a expected way, not a error.
I suggest change those log level to info.

##### Expected behavior

check failed modules should report message as not error but info.

**Comment by gmosx** [comment]
Sounds reasonable to me.

**Comment by Wing924** [comment]
@gmosx not for just python.d but for all collectors
```
2019-02-23 14:56:40: netdata ERROR : PLUGIN[proc] : BTRFS: Cannot open directory '/sys/fs/btrfs'. (er
rno 2, No such file or directory)
2019-02-23 05:56:41: go.d ERROR: main[main] logstash[local] Check failed
2019-02-23 14:56:41: fping.plugin: FATAL: no hosts configured - nothing to do.
```

**Comment by Ferroin** [comment]
If the user has explicitly enabled the collector or module, this message should _probably_ be an error, because the user is expecting data collection for the collector or module, but if it's just been enabled by default by Netdata, then using INFO for the level makes sense.

**Comment by ilyam8** [comment]
> If the user has explicitly enabled the collector or module, this message should _probably_ be an error,

Totally agree. In context of job this is certainly `error`.

**Comment by ilyam8** [comment]
> If servers don't have nginx

We don't aware of that, if there is a job, then we expect that it should run.

But in general i agree, for example this
https://github.com/netdata/netdata/blob/2ee2d94891e88d9c773f7633746a38fb341f2bac/collectors/python.d.plugin/python.d.plugin.in#L359

shouldn't be and error (because it is not error).

**Comment by Wing924** [comment]
> We don't distinguish between user defined and default jobs, so `error` is the right level.

I disagree with it. 

* The info level is `a notice for users`.
* The error level is `something bad happen and users need to take some action`.

Because Netdata is designed with auto discovery everything, we should allow discovery failed. The message should be `info` level. 

Another point is too many error equals no error. Users tend to ignore errors if almost of them are false positive.

**Comment by ilyam8** [comment]
> the user is expecting data collection for the collector or module

> web_log: nginx_log: /var/log/nginx/access.log not readable or not exist

The message is from job and the level is definitly not `info`, it is `error` and it requires user intervention.

This is true for user defined jobs, but we treat all jobs as user defined for now.

**Comment by Wing924** [comment]
> This is true for user defined jobs, but we treat all jobs as user defined for now.

If netdata's config defaults to empty, it should be error. But most of them are pre-defined job, so that we should not treat them as error. The error message is toooooooooo noisy which lead user not to notice real important errors, such as segfault.

**Comment by ilyam8** [comment]
There is no such thing as predefined/user jobs, all of them are just jobs. 

To add new job user has 2 options:
 - add new job
 - modify predefined job

I, for example, quite ofthen use opt 2.

Since all of them are jobs we should treat them equaly. It will be frustrating if one job dies with `info` level and another with `error`.

And again, in context of job, it is not `info`, it is `error`.

> which lead user not to notice real important errors, such as segfault.

I understand the gist and agree in general, but i don't think that moving all errors to info level is the way to go.

**Comment by Wing924** [comment]
> It will be frustrating if one job dies with `info` level and another with `error`.

I suggest treat both to be `info`.

I think the UX is much more important than philosophy about correct log level.
May be we need to introduce another log file to log check failed.

For a light user who just run starter script, they may be scared by so many errors.
For heavy user like me, just ignored segfault for a few month.

**Comment by ilyam8** [comment]
Ok, so the problem is not incorrect log level, but a mess in the `error.log`.

@cakrit @ktsaou 

What about to have several `error.log` files? For example we could have a separate log file for external plugins.

**Comment by cakrit** [comment]
My opinion:

ERROR is a call to action. Any system's normal operation should result in 0 errors. Whenever we can't distinguish between a true error condition and an info, it's safe to be on the ERROR side, but we have greatly overdone it. 
- We do not require users to create a user config, yet we log an ERROR when it's not there. This must be converted to info, if logged at all. Makes no sense as an error.
- Some collectors can automatically identify whether a service is available, even without a user config. In these cases, if I don't find what I'm looking for, it's clearly NOT an error. The only way it could possibly be an error, is if that thing we're looking for should exist in ALL systems, regardless of OS, configuration etc. 

For us, would only log an ERROR if I know for a fact that a user requested something to be monitored and I'm not able to do it. The only indication for this scenario is the existence a user config file. 

In short: 
- No user configuration for a particular collector? No errors shown, unless we're 100% certain that what is monitored should exist in all installations.
- User configuration exists? Sure, raise the log level for that collector, as the user expects to see it.

**Comment by ilyam8** [comment]
> To add new job user has 2 options:
>
> * add new job
> * modify predefined job

> I, for example, quite ofthen use opt 2.


> For us, would only log an ERROR if I know for a fact that a user requested something to be monitored and I'm not able to do it.
> a fact that a user requested something to be monitored


How? 

These are predefined jobs:

https://github.com/netdata/netdata/blob/88d191cf32f8bdcbb40e370325b3e6f3d9971c31/collectors/python.d.plugin/apache/apache.conf#L75-L85

User is not supposed to change them in most of the cases. User enables status report and that is all.

So suggested login doesn't work in that case.

**Comment by Ferroin** [comment]
I would tend to agree with @cakrit on this.

Looking at it a bit differently, if I've explicitly configured collector X to monitor service Y, and it can't find service Y, it's an error, period, either something is wrong in my configuration (not necessarily in Netdata but _somewhere_), or there is a bug somewhere in the system (or a hardware error, but the first two cases are much more likely).

I can entirely understand wanting there to not be so many errors in the log, but just blindly removing things that are obviously errors is going to potentially cause problems in the other direction.

A bit more concretely, I'd argue for the following logic here:

* If the service to be monitored has successfully had data collected from it at least once, any errors collecting further data should be reported as errors because things suddenly not working when they had been previously are pretty universally accepted as errors.
* If the service has not successfully had any data collected from it, and the user has not explicitly configured things (that is, there is no config for this collection in `/etc/netdata`), treat any errors other than the collector/module crashing as INFO messages, not errors.
* If the service has not successfully had any data collected from it, and the user has explicitly configured things (that is, there is config for this collection in `/etc/netdata`), treat any errors as errors.

**Comment by Ferroin** [comment]
@ilyam8 Why not just look at whether the config is being pulled from `/etc/netdata` instead of wherever the default configs are stored on the system?

**Comment by ilyam8** [comment]
Guys, what about having a separate `error.log` and don't add all that fancy if-else-if logic?

**Comment by ilyam8** [comment]
> * and the user has not explicitly configured things (that is, there is no config for this collection in `/etc/netdata`)

for example, you don't need to have anything in `/etc/netdata` for apache module

**Comment by Wing924** [comment]
> I can entirely understand wanting there to not be so many errors in the log, but just blindly removing things that are obviously errors is going to potentially cause problems in the other direction.

We have both error log and info log, right? So we won't lose log, just change the output file.

> If the service to be monitored has successfully had data collected from it at least once, any errors collecting further data should be reported as errors because things suddenly not working when they had been previously are pretty universally accepted as errors.

I agree with it.

> If the service has not successfully had any data collected from it, and the user has not explicitly configured things (that is, there is no config for this collection in `/etc/netdata`), treat any errors other than the collector/module crashing as INFO messages, not errors.

As netdata is designed to have more job than actually need to have,  explicitly configured things also may contain jobs with check failed as normal case. For example, because I manage 100k+ instances and I can't config each server by hand, I put same springboot config on all servers and tell application developers: please run springboot on port 8080 so that we can monitor that.

**Comment by Ferroin** [comment]
> for example, you don't need to have anything in `/etc/netdata` for apache module

Yes, but that's only if you are just using the default config.  My point is that the default config shouldn't throw errors when it fails to detect something, but a user specified config probably should.

> We have both error log and info log, right? So we won't lose log, just change the output file.

Except that failing to collect data that the user has explicitly asked for us to collect is an error, plain and simple.  Logging it as info is misleading at best, and at worst will cause the same kind of issue that you are trying to solve with this change (namely, the user won't notice the message).  And, unless I'm significantly misunderstanding how the alert handling works, we can't use an alert to check if something is being monitored or not if it was never successfully monitored in the first place, so we kind of need to log this sanely.

>As netdata is designed to have more job than actually need to have, explicitly configured things also may contain jobs with check failed as normal case. For example, because I manage 100k+ instances and I can't config each server by hand, I put same springboot config on all servers and tell application developers: please run springboot on port 8080 so that we can monitor that.

Perhaps we should just make a config option that controls whether failure to start monitoring a service constitutes an error or not?  I mean, ideally we would have some way to see if what we are trying to monitor is even there, and only log an error if it appears to be there but we can't get data, but that's likely to get _really_ complicated to implement properly.

**Comment by ilyam8** [comment]
> We have both error log and info log, right? So we won't lose log, just change the output file.

we don't

**Comment by Wing924** [comment]
Simply change to `info` level and log that in `info.log` file is hard?

**Comment by ilyam8** [comment]
it will be hard to track logging for a module if it logs to several places. (`grep module_name error.log` won't work)

I disagree that it should be info, just because _we have a lot of errors in error.log_.

**Comment by Wing924** [comment]
> it will be hard to track logging for a module if it logs to several places. (`grep module_name error.log` won't work)

`grep module_name *.log`

**Comment by ilyam8** [comment]
log messages order is a thing! :smile:

**Comment by Wing924** [comment]
> I disagree that it should be info, just because _we have a lot of errors in error.log_.

The `error.log` have all logs including `info`, `warning` and `error`. 
How about change to `warning`? Heavy users can use `grep error.log -v WARNING` to ignore those annoy messages, but still provide info of something may be wrong to all users.

The big thing is we should distinguish `may be wrong` and `must be wrong`.

**Comment by cakrit** [comment]
I voted for @Ferroin 's proposal earlier which is an improvement over mine. It's as complex as it needs to be, given the circumstances. 

The more log files one adds, the more difficult it is to discover what's wrong. Good luck ordering log entries from multiple files to see the series of events, especially when you get entries at the same second.

**Comment by Ansem93** [comment]
My personal opinion after I was scared but the 300kb of errors for missing sensors.
I think that treat this as an ERROR is a mistake as many saids. Mostly because 99% of the time it's just normal that this sensor is missing.
I also think that it's wrong to label it as INFO since INFO is normal behavoir. Not findind a sensor is not normal from the Netdata point of view.

I would propose to things:
1) Laber it as WARNING: it's not an info so a user is not scared thinking netdata is broken, but at least you know that it may be an issue
2) Change the log wording. Writing something like 
"WARNING: MISSING SENSOR: IPv6 bind() on ip '::1' port 8125, socktype 2 failed. (errno 99, Address not available)
"WARNING: MISSING SENSOR: Cannot bind to ip '::1', port 8125"
Instead of
"2020-02-20 12:12:48: netdata ERROR : STATSD : LISTENER: IPv6 bind() on ip '::1' port 8125, socktype 2 failed. (errno 99, Address not available)
2020-02-20 12:12:48: netdata ERROR : STATSD : LISTENER: Cannot bind to ip '::1', port 8125"

**Comment by ilyam8** [comment]
Closed in favour of #6038
